File: cpp/day9/main.cpp
#include "utils/utils.h"
#include <cctype>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <limits>
#include <map>
#include <optional>
#include <set>
#include <sstream>
#include <string>
#include <vector>

const uint16_t EMPTY = std::numeric_limits<uint16_t>::max();

std::vector<uint16_t> expand_disk(std::string_view compacted) {
    std::vector<uint16_t> disk;
    for (size_t i = 0; i < compacted.length(); i++) {
        if (!isdigit(compacted[i])) {
            continue;
        }

        size_t len = compacted[i] - '0';
        if (i % 2 == 0) {
            disk.insert(disk.end(), len, i / 2);
        }
        else {
            disk.insert(disk.end(), len, EMPTY);
        }
    }

    return disk;
}

#ifndef AOC_PART2
void frag(std::vector<uint16_t> &disk) {
    size_t i = 0;
    size_t j = disk.size() - 1;

    while (true) {
        while (i < disk.size() && disk[i] != EMPTY) {
            i++;
        }

        while (i < j && disk[j] == EMPTY) {
            j--;
        }

        std::swap(disk[i], disk[j]);
        i++;

        if (j == 0 || i >= j) {
            break;
        }

        j--;
    }
}
#else
std::tuple<std::vector<size_t>, std::map<size_t,std::set<size_t>>> get_disk_map(std::string_view compacted) {
    std::vector<size_t> file_lengths;
    std::map<size_t, std::set<size_t>> free_space;
    size_t acc = 0;

    for (size_t i = 0; i < compacted.length(); i++) {
        if (!isdigit(compacted[i])) {
            continue;
        }

        size_t length = static_cast<size_t>(compacted[i] - '0');
        if (i % 2 == 0) {
            file_lengths.push_back(length);
        }
        else {
            free_space[length].insert(acc);
        }
        acc += length;
    }

    return std::make_tuple(file_lengths, free_space);
}

void defrag(std::vector<uint16_t> &disk, const std::vector<size_t> &file_lengths, std::map<size_t, std::set<size_t>> &free_space) {
    size_t j = disk.size() - 1;

    while (true) {
        while (disk[j] == EMPTY) {
            if (j == 0)
                return;
            j--;
        }

        size_t file_id = disk[j];
        size_t file_length = file_lengths[file_id];

        size_t space_index = j;
        size_t space_amount = 0;
        auto space = free_space.lower_bound(file_length);
        while (space != free_space.end()) {
            auto first_available = space->second.begin();
            if (first_available != space->second.end() && *first_available < space_index) {
                space_index = *first_available;
                space_amount = space->first;
            }
            space++;
        }

        if (space_index == j) {
            // couldn't find a suitable spot
            if (j <= file_length)
                return;
            j -= file_length;
            continue;
        }

        for (size_t i = space_index; i < space_index + file_length; i++, j--) {
            std::swap(disk[i], disk[j]);
        }

        // bookkeeping
        free_space[space_amount].erase(space_index);
        if (file_length < space_amount) {
            free_space[space_amount - file_length].insert(space_index + file_length);
        }
    }
}
#endif

uint64_t checksum(const std::vector<uint16_t> &disk) {
    uint64_t checksum = 0;
    size_t i = 0;

    while (i < disk.size()) {
        if (disk[i] != EMPTY) {
            checksum += static_cast<uint64_t>(disk[i]) * i;
        }
        i++;
    }

    return checksum;
}

int main(int argc, char *argv[]) {
    Timer t;

    if (argc < 2) {
        std::cerr << "Not enough arguments" << std::endl;
        return 1;
    }

    std::ifstream input(argv[1]);
    std::stringstream buffer;
    buffer << input.rdbuf();  

    auto disk = expand_disk(buffer.str());
#ifndef AOC_PART2
    frag(disk);
#else
    std::vector<size_t> file_lengths;
    std::map<size_t, std::set<size_t>> free_space;
    std::tie(file_lengths, free_space) = get_disk_map(buffer.str());
    defrag(disk, file_lengths, free_space);
#endif
    auto answer = checksum(disk);

    std::cout << "Answer: " << answer << std::endl;
}

File: rs/day9/src/main.rs
use utils::timer::Timer;
use std::collections::{BTreeMap, BTreeSet};
use std::fs::read;
use std::path::PathBuf;

const EMPTY: u16 = u16::MAX;

fn expand_disk(compacted_disk: &[u8]) -> Vec<u16> {
    compacted_disk
        .iter()
        .enumerate()
        .filter(|(_, ch)| ch.is_ascii_digit())
        .flat_map(|(idx, ch)| {
            (0..(ch - b'0'))
                .map(move |_| {
                    if idx % 2 == 1 { 
                        EMPTY
                    }
                    else {
                        (idx / 2) as u16
                    }
                })
        })
        .collect::<Vec<u16>>()
}

fn frag(disk: &mut [u16]) {
    let mut i = 0;
    let mut j = disk.len() - 1;

    loop {
        while i < disk.len() && disk[i] != EMPTY {
            i += 1;
        }
        while i < j && disk[j] == EMPTY {
            j -= 1;
        }

        disk.swap(i, j);
        i += 1;
        
        if j == 0 || i >= j {
            break;
        }
        
        j -= 1;
    }
}

fn get_disk_map(compacted_disk: &[u8]) -> (Vec<usize>, BTreeMap<usize, BTreeSet<usize>>) {
    let (file_lengths, free_space, _) = compacted_disk
        .iter()
        .enumerate()
        .filter(|(_, ch)| ch.is_ascii_digit())
        .map(|(idx, &ch)| (idx, (ch - b'0') as usize))
        .fold((Vec::new(), BTreeMap::new(), 0usize), |(mut file_lengths, mut free_space, acc), (idx, len)| {
            if idx % 2 == 1 {
                free_space.entry(len).or_insert(BTreeSet::new()).insert(acc);
            }
            else {
                file_lengths.push(len);
            }
            (file_lengths, free_space, acc + len)
        });
    (file_lengths, free_space)
}

fn defrag(disk: &mut [u16], file_lengths: &[usize], mut free_spaces: BTreeMap<usize, BTreeSet<usize>>) {
    let mut j = disk.len() - 1;

    'outer: loop {
        while disk[j] == EMPTY {
            j -= 1;
            if j == 0 {
                break 'outer;
            }
        }

        let file_id = disk[j];
        let file_length = file_lengths[file_id as usize];

        let (space_length, space_index) = {
            let Some((space_length, space_index)) = free_spaces
                .range(file_length..)
                .map(|(space_size, spaces)| (space_size, spaces.first()))
                .filter(|(_, index)| index.is_some_and(|&f| f < j))
                .map(|(space_length, index)| (space_length, index.unwrap()))
                .min_by_key(|(_, index)| **index) else {
                if j < file_length {
                    break 'outer;
                }
                j -= file_length;
                continue;
            };

            (*space_length, *space_index)
        };

        if space_index > j {
            if j < file_length {
                break 'outer;
            }
            j -= file_length;
            continue;
        }
                
        for i in space_index..space_index+file_length {
            disk.swap(i, j);
            j -= 1;
        }
        
        // bookkeeping
        let space_left = space_length - file_length;
        free_spaces.entry(space_length).and_modify(|indices| { indices.remove(&space_index); });
        if space_left > 0 {
            free_spaces.entry(space_left).or_default().insert(space_index + file_length);
        }
    }
}

fn checksum(disk: &[u16]) -> usize {
    disk
        .iter()
        .enumerate()
        .filter(|(_, &d)| d != EMPTY)
        .map(|(idx, &d)| idx * d as usize)
        .sum()
}

fn main() {
    let _timer = Timer::new();

    let path: PathBuf = std::env::args_os()
        .nth(1)
        .expect("Should have file argument")
        .into();

    let disk = read(path)
        .expect("Should be able to read from path");

    let mut expanded = expand_disk(&disk);
    
    if cfg!(feature = "part2") {
        let (file_lengths, free_space) = get_disk_map(&disk);
        defrag(&mut expanded, &file_lengths, free_space);
    }
    else {
        frag(&mut expanded);
    }

    let answer = checksum(&expanded);

    println!("Answer: {answer}");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_expand() {
        let example = b"12345";
        let expanded = expand_disk(example);

        assert_eq!(expanded,
            vec![0, EMPTY, EMPTY, 1, 1, 1, EMPTY, EMPTY, EMPTY, EMPTY, 2, 2, 2, 2, 2]);
    }

    #[test]
    fn test_defrag() {
        let example = b"2333133121414131402";
        let mut expanded = expand_disk(example);
        let (file_lengths, free_space) = get_disk_map(example);
        defrag(&mut expanded, &file_lengths, free_space);
        let checksum = checksum(&expanded);
        assert_eq!(checksum, 2858);
    }
}
